src/OpenLoco/MessageManager.cpp

```diff
@@ -37,9 +40,74 @@ namespace OpenLoco::MessageManager
         call(0x004285BA, regs);
     }
 
+    // 0x00428E0F
+    static void clearActiveMessage()
+    {
+        if (getGameState().activeMessageIndex != MessageId::null)
+        {
+            get(getGameState().activeMessageIndex)->var_C8 = 0xFFFF;
+        }
+        Ui::WindowManager::close(Ui::WindowType::news);
+        getGameState().activeMessageIndex = MessageId::null;
+    }
+
+    // 0x00428DAB
+    static void remove(const MessageId id)
+    {
+        if (getGameState().activeMessageIndex != MessageId::null)
+        {
+            if (id == getGameState().activeMessageIndex)
+            {
+                clearActiveMessage();
+            }
+            else if (enumValue(id) < enumValue(getGameState().activeMessageIndex))
+            {
+                getGameState().activeMessageIndex = static_cast<MessageId>(enumValue(getGameState().activeMessageIndex) - 1);
+            }
+        }
+        numMessages()--;
+        // Move element to end of array
+        if (enumValue(id) < Limits::kMaxMessages - 1)
+        {
+            std::rotate(get(id), get(id) + 1, std::end(rawMessages()));
+        }
+        Ui::WindowManager::invalidate(Ui::WindowType::messages);
+    }
+
+    // 0x0042851C
+    void removeAllSubjectRefs(const uint16_t subject, MessageItemArgumentType type)
+    {
+        for (auto i = numMessages(); i > 0; --i)
+        {
+            auto& message = rawMessages()[i];
+            const auto& descriptor = getMessageTypeDescriptor(message.type);
+            bool hasModified = false;
+            for (auto j = 0; j < 3; ++j)
+            {
+                if (descriptor.argumentTypes[j] == type
+                    && message.itemSubjects[j] == subject)
+                {
+                    message.itemSubjects[j] = 0xFFFF;
+                    hasModified = true;
+                }
+            }
+            if (hasModified && getGameState().activeMessageIndex == static_cast<MessageId>(i))
+            {
+                Ui::WindowManager::invalidate(Ui::WindowType::news);
+            }
+        }
+    }
+
     // 0x004284DB
     void updateDaily()
     {
-        call(0x004284DB);
+        for (auto i = numMessages() - 1; i >= 0; --i)
```

Comment:
if you change `GameState::messages` from c-style array into std::array (`std::array<Message, Limits::kMaxMessages> messages;`), these loops become much nicer range-based for loops.

Reply:
Sadly not possible due to the size being defined by the numMessages